domain/timeline/src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/domain/timeline/service/TimelineService.kt

```diff
@@ -16,6 +17,23 @@ class TimelineService(
   private val persistenceAdapter: TimelinePersistenceAdapter,
 ) {
 
+  /**
+   * Records an [TimelineEvent] that has taken place for a prisoner.
+   */
+  fun recordTimelineEvent(prisonNumber: String, event: TimelineEvent) {
+    persistenceAdapter.recordTimelineEvent(prisonNumber, event)
+  }
+
+  /**
+   * Records a collection of [TimelineEvent]s that have taken place for a prisoner.
+   */
+  fun recordTimelineEvents(prisonNumber: String, events: List<TimelineEvent>) {
+    // TODO RR-314 - write test
+    events.forEach {
```

Comment:
Just thinking out load here, and I could be wrong ....
But I wonder if the loop should in the persistence adapter method impl, rather than here in the service?
IE. have a method in the persistence adapter that takes the collection. The implementation would be annotated `@Transactional` and would do the looping (or perhaps a `saveAll` type method call?)

Anyway, I guess my worry is about TX boundaries. As you have it here the TX boundary would be around each save, so you might end up with some of the events saved, but not all of them.
If we do it as I've suggested the TX boundary would be around all of the timeline events, so if one fail, they all rollback 👍